modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java

```diff
@@ -2156,75 +2153,53 @@ public Map<String, Map<String, String>> getWorkflowStateMappings() {
 
   @Override
   public void repopulate(final ElasticsearchIndex index) throws IndexRebuildException {
-    final String startWorkflow = Operation.START_WORKFLOW.toString();
     final int total;
     try {
-      total = serviceRegistry.getJobCount(startWorkflow);
-    } catch (ServiceRegistryException e) {
+      total = persistence.countMediaPackages();
+    } catch (WorkflowDatabaseException e) {
       logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
       throw new IndexRebuildException(index.getIndexName(), getService(), e);
     }
-    final int limit = 1000;
 
     if (total > 0) {
       logIndexRebuildBegin(logger.getSlf4jLogger(), index.getIndexName(), total, "workflows");
+      List<WorkflowIndexData> workflowIndexData;
+
+      try {
+        workflowIndexData = persistence.getWorkflowIndexData();
+      } catch (WorkflowDatabaseException e) {
+        logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
+        throw new IndexRebuildException(index.getIndexName(), getService(), e);
+      }
+
+      Map<String, WorkflowIndexData> latestWorkflows = new HashMap<>();
+      for (WorkflowIndexData data : workflowIndexData) {
+        latestWorkflows.put(data.getMediaPackageId(), data);
+      }
+
       int current = 0;
-      int offset = 0;
-      List<WorkflowInstance> workflows;
       do {
-        try {
-          workflows = persistence.getWorkflowInstances(limit, offset);
-        } catch (Exception e) {
-          logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), total, current, e);
-          throw new IndexRebuildException(index.getIndexName(), getService(), e);
-        }
-        logger.debug("Got {} workflows for re-indexing", workflows.size());
-        offset += limit;
-
-        for (WorkflowInstance instance : workflows) {
+        logger.debug("Got {} workflows for re-indexing", latestWorkflows.size());
+        for (WorkflowIndexData data : latestWorkflows.values()) {
           current += 1;
           Organization organization = null;
           try {
-            organization = organizationDirectoryService.getOrganization(instance.getOrganizationId());
+            organization = organizationDirectoryService.getOrganization(data.getOrganizationId());
           } catch (NotFoundException e) {
-            logger.error("Found workflow with non-existing organization {}", instance.getOrganizationId());
+            logger.error("Found workflow with non-existing organization {}", data.getOrganizationId());
             continue;
           }
```

Comment:
If we don't need to switch the security context below, we also wouldn't need to retrieve the organization here but could just pass on the organization identifier.